rlbot_gui/story/story_challenge_setup.py

```diff
@@ -0,0 +1,462 @@
+        time.sleep(0.5)
+        setup_manager.game_interface.fresh_live_data_packet(packet, 1000, WITNESS_ID)
+
+    if packet.num_cars == 0:
+        print("The game was initialized with no cars")
+        return early_failure
+
+    tick_rate = 120
+    results = None
+    max_boost = 0
+    if "boost-100" in upgrades:
+        max_boost = 100
+    elif "boost-33" in upgrades:
+        max_boost = 33
+
+    half_field = challenge.get("limitations", []).count("half-field") > 0
+
+    stats_tracker = ManualStatsTracker(challenge)
+    last_boost_bump_time = time.monotonic()
+    while True:
+        try:
+            packet = GameTickPacket()
+            setup_manager.game_interface.fresh_live_data_packet(
+                packet, 1000, WITNESS_ID
+            )
+
+            if packet.num_cars == 0:
+                # User seems to have ended the match
+                print("User ended the match")
+                return early_failure
+
+            stats_tracker.updateStats(packet)
+            results = packet_to_game_results(packet)
+
+            if has_user_perma_failed(challenge, stats_tracker.stats):
+                time.sleep(3)
+                setup_failure_freeplay(setup_manager, "You failed the challenge!")
+                return early_failure
+            
+            if end_by_mercy(challenge, stats_tracker.stats, results):
+                time.sleep(5)
+                setup_failure_freeplay(setup_manager, "Challenge completed by mercy rule!", "green")
+                return True, results
+
+            game_state = GameState.create_from_gametickpacket(packet)
+            # game_state.console_commands.append("ShowDebug PHYSICS")
+            human_state = game_state.cars[0]  # this is used to change state
+
+            changed = False
+            # adjust boost
+            if human_state.boost_amount > max_boost and not half_field:
+                # Adjust boost, unless in heatseeker mode
+                human_state.boost_amount = max_boost
+                changed = True
+
+            if "boost-recharge" in upgrades:
+                # increase boost at 10% per second
+                # we do it in chunks because otherwise we lag hard
+                now = time.monotonic()
+                if human_state.boost_amount < max_boost and (now - last_boost_bump_time > 2):
+                    changed = True
+                    last_boost_bump_time = now
+                    human_state.boost_amount += 20
+
+
+            # This isn't fun :)
+            # if half_field and human_state.physics.location.y > 0:
+            #     human_state.physics.location.y = 0
+            #     human_state.physics.velocity.y = -abs(human_state.physics.velocity.y)
+            #     changed = True
+
+            if changed:
+                setup_manager.game_interface.set_game_state(game_state)
+
+            if packet.game_info.is_match_ended:
+                break
+
+            time.sleep(1.0 / tick_rate)
```

Comment:
This is slightly at odds with your use of game_interface.fresh_live_data_packet, which blocks until a new packet is available. Theoretically you could get rid of the sleep and automatically loop at whatever tick rate falls out of this: https://github.com/RLBot/RLBot/wiki/Tick-Rate

If you prefer to control your tick rate manually, update_live_data_packet returns immediately and may be preferable.

Reply:
Ah thanks, have gone with just processing the ticks as they arrive.